Refactor the post checkout code

niranjan-uma-shankar opened the pull request:
#### Changes proposed in this Pull Request

* Make the fallback  destination as Thank You page if there's a site slug and item in cart, else `/`.
* Removes a bunch of code which is no longer needed in `client/state/selectors/is-eligible-for-signup-destination.js`. 
    - removes a few redundant checks. We no longer need to check each cart item and return the Thank You page for non-plan products, since the default fallback destination is now the Thank you page.
    - we no longer need the check for `isNewSite`, which was used to determine if we should show the Thank You page for in-site purchases. As said in the previous bullet point, every purchase made outside of a signup flow will now always show the Thank You page.

#### Fixes
https://github.com/Automattic/wp-calypso/issues/35068 and  https://github.com/Automattic/wp-calypso/issues/35284

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Scenario 1**

* Begin at http://calypso.localhost:3000/start and complete the signup flow with a plan purchase. 
* Post checkout you will be shown a Quick Start session upsell.
* Clicking Yes(and completing purchase) or No should both take to signup destination,
which is checklist for most flows, and thank you page if you have an eCommerce plan in the cart. Verify with/without eCommerce plan.


**Scenario 2**

* On an existing account, add new site and purchase a paid plan. Verify that you are taken to the appropriate signup destination(which is checklist for most flows, and thank you page if you have an eCommerce plan in the cart).

**Scenario 3**(private by default launch flow)

* Verify that https://github.com/Automattic/wp-calypso/issues/35284 is fixed.

**Scenario 4**(testing in-site purchases)

* On an existing new account(less than 30 minutes old), upgrade your plan to the next tier.
* Post checkout, you should be taken to the Thank You page.
* Purchase other products, such as a domain or theme. Verify that you are taken to the Thank You page post checkout.(this verifies that we no longer need the `isNewSite` check as described in the PR description)

**Scenario 5**

* Go through the /start/business flow and add a custom domain to cart. Complete purchase.
* You will not be shown the Quick Start nudge for business plan purchase, however you will see a GSuite upsell(in sandbox, GSuite can be purchased only for .blog domains. Check FG for more details).
* Verify that the final destination is the checklist.

**Scenario 6**(Jetpack tests)

_Jetpack connect_

* Spin up a Jetpack connected instance at https://jurassic.ninja
* Go through the setup flow (note that you will get redirect to a wordpress.com URL when you click the Set up Jetpack button in wp-admin, so ensure that you stay on calypso.localhost URL).
* Verify that after purchasing a Jetpack plan, you are taken to my-plans page /plans/my-plan/SITE_SLUG?thank-you=&install=all

_Jetpack upgrade_

Verify that https://github.com/Automattic/wp-calypso/issues/35068 is fixed.

**Scenario 7**

* Begin at /start and purchase a plan
* Pay with a redirect payment type such as Giropay(proxy through Europe) or PayPal
* Verify that post purchase, you are taken to the checkout pending page(destination is not respected for redirected payment types for the moment, and is tracked in https://github.com/Automattic/wp-calypso/issues/34333.

**Scenario 8**

* Begin at http://calypso.localhost:3000/themes and select to purchase any Premium plan.
* Go through the `with-theme` signup flow and complete checkout - try both with a free plan and a paid plan.
* In both scenarios(free plan or paid plan), verify that you are taken to the checklist page with a theme specific sub-header text:
<img width="981" alt="Screenshot 2019-08-20 at 5 23 03 PM" src="https://user-images.githubusercontent.com/1269602/63351334-a9953600-c37c-11e9-87dc-6b6ee7bae18d.png">

niranjan-uma-shankar committed:
Refactor the post checkout code

niranjan-uma-shankar committed:
Fallback destination should be the thank you page if cart is not empty, even if there's no receipt number

niranjan-uma-shankar committed:
Minor: spacing

niranjan-uma-shankar committed:
clear the cookie if there's an ecommerce plan in cart

spen commented:
In case anybody starts reviewing while I'm part way through, so far I've tested:
- [x] Scenario 1 ✅ 
- [x] Scenario 2 ✅ 
- [x] Scenario 3 ✅ 
- [x] Scenario 4 ✅  
- [ ] Scenario 5 
- [ ] Scenario 6 
- [ ] Scenario 7

spen commented:
Hey, I'm struggling with the Jetpack testing. 
Here's what I got stuck on:

> Verify that after purchasing a Jetpack plan, you are taken to my-plans page /plans/my-plan/SITE_SLUG?thank-you

I ended up at `/plans/my-plan/possible-jellyfish.jurassic.ninja?install=all`:

<img width="1680" alt="Screenshot 2019-08-13 at 23 18 04" src="https://user-images.githubusercontent.com/4335450/62981320-d39ca480-be20-11e9-8cd3-3162c854a5e3.png">

I followed the instructions there, going through JN and replacing `wordpress.com/...` with `calypso.localhost:3000`. I tried this a few times thinking I'd maybe made mistakes etc. but I'd either see network issues (I think my service was acting up) or I'd end up here at `/plans/my-plan/`

niranjan-uma-shankar commented:
Going through Setup Jetpack flow, if you land at `/plans/my-plan/possible-jellyfish.jurassic.ninja?thank-you=&install=all`, then that's the correct destination. I will update testing instruction to include the `install=all` attribute. Thanks for testing everything.

niranjan-uma-shankar committed:
Missing arg for the this.maybeRedirectToGSuiteNudge() function

niranjan-uma-shankar commented:
Pushed in a small fix and updated Scenario 5 test case to cover this.

spen commented:
I finally managed to get set up to test the last test scenario and it is as described in the issue

spen reviewed (approved):
I've triggered rebuilds for the failing tests there, but if those come back ✅ then let's merge! 👍 
Nice work and thanks a bunch for all the details and help :)

niranjan-uma-shankar commented:
We have a failing test for the themes signup flow, and we need to decide the behaviour:

1. On production, if you go via the `with-theme` signup flow(by selecting to purchase a premium theme in https://wordpress.com/themes/premium while logged out), then selecting a free plan takes to a theme specific thank you page([check screenshot](https://cloudup.com/ceoB3BDsxxo)) whereas selecting a paid plan in the flow takes to the [flow destination].(https://github.com/Automattic/wp-calypso/blob/master/client/signup/config/flows-pure.js#L114).
2. The good news is that this PR fixes the issue by always taking to flow destination for either a free or paid plan in this flow. Should we fix the e2e test then? Or, is there a reason to show the theme specific thank you page in this flow for free plans?

spen commented:
> Or, is there a reason to show the theme specific thank you page in this flow for free plans?

Giving it some thought I think we should go with the theme specific thank you page - it gives credit to the theme authors and lets the customer get straight in to learning about the theme features and so on with the CTAs in the modal.

In terms of behaviour, I'm not 100% sure even after giving this some thought.
If we're already logged in and we purchase a theme it totally makes sense to be landed on the theme specific thankyou page - the theme is all that's on our mind so it makes sense to then be presented with the options to either "learn more about this theme" or "customize site".

When going to `/themes` as a logged out user that's a different story. This user was primarily focused on exploring a few themes, then picked one and, while making their way through the `with-theme` flow, created an account.

It doesn't make sense to me for the first directions to show after creating an account are "learn more about this theme" or "customize site" - I'd rather see the checklist, or a preview of my new site.

That said, I'd still like to see credit go to the theme author in some way, I'm just not sure how or where would be better than the current modal.

Personally I'd err on the side of taking folk to the regular sign up destination when the customer has just created a new account. Else we may be dropping them in to part of WP.com that's not as new-user-friendly.

niranjan-uma-shankar commented:
Fully agree @spen . Then I'll fix the e2e test to reflect this.

Do you think it makes sense to show something related to the theme purchase in the sub-header? For example, if you purchase GSuite in the signup flow, the checklist sub-header shows this message:

<img width="867" alt="Screenshot 2019-08-16 at 7 47 22 PM" src="https://user-images.githubusercontent.com/1269602/63174133-fb794b80-c05e-11e9-89ef-0e36119c6111.png">

I modelled a similar sub-header if a concierge session is purchased in the signup flow:

<img width="956" alt="Screenshot 2019-08-16 at 7 47 06 PM" src="https://user-images.githubusercontent.com/1269602/63174172-0f24b200-c05f-11e9-8164-2f0e140c9375.png">

Could we show something on the theme purchase? And maybe a link to the "learn more about this theme" in the sub-header?

spen commented:
I really like that idea! It'd be extra cool if it could mention the theme name & author! Let's go for it :)

niranjan-uma-shankar commented:
I'll implement that then :)

simison commented:
>Going through Setup Jetpack flow, if you land at /plans/my-plan/possible-jellyfish.jurassic.ninja?thank-you=&install=all, then that's the correct destination. I will update testing instruction to include the install=all attribute. Thanks for testing everything.

Note that once https://github.com/Automattic/wp-calypso/pull/35471 merges, `&install=all` is no more expected. Thanks for testing Jetpack flows!

niranjan-uma-shankar committed:
Theme signup flow changes
In the with-theme signup flow, the destination will always be checklist. The checklist page will show a snippet about the theme name and author in the subheader. E2E test has been fixed since it expects the the destination for with-theme signup flow to be the theme thanks modal for a free plan in the cart.

niranjan-uma-shankar committed:
Updating modified date for the with-theme flow definition

niranjan-uma-shankar committed:
Remove Jetpack redirect from thank you page, since it's handled in post checkout redirect

niranjan-uma-shankar commented:
1. In [c1092b2](https://github.com/Automattic/wp-calypso/pull/35283/commits/c1092b28117e591d9422a05a3cb93e8634b09d19), I've added a theme specific sub-header text while going through the `with-theme` flow. 

<img width="981" alt="Screenshot 2019-08-20 at 5 23 03 PM" src="https://user-images.githubusercontent.com/1269602/63351334-a9953600-c37c-11e9-87dc-6b6ee7bae18d.png">

2. I've removed a redundant piece of code for jetpack redirection in [244e6a0](https://github.com/Automattic/wp-calypso/pull/35283/commits/244e6a0a6dac9ab2644eca184cd6ef754039b277), since we're handling that already [here](https://github.com/Automattic/wp-calypso/pull/35283/files#diff-199e3967ae413c2ff48db966abbf8386R396).

@spen This is ready for another review again. Thanks!

niranjan-uma-shankar committed:
some minor fixes to make the variables const

niranjan-uma-shankar commented:
I'll just leave the `In Pogress` label on for a while till the e2e tests pass. It makes sense to continue the review when we have all greens.

niranjan-uma-shankar commented:
After rebase, the `with-theme` flow destination is not getting the `?d=theme` query param(which is needed to display the theme specific sub-header in the checklist page). This is because https://github.com/Automattic/wp-calypso/pull/35392 adds a redirect from /checklist to /home, and in the process, does not carry forward query params in /checklist to /home. 

Until that gets fixed, we needs to test the `with-theme` flow by turning off the feature flag for customer home. So, go through the with-theme flow, and in the checkout page, modify the browser url to `/checkout/{YOUR_SITE_SLUG}?flags=-customer-home` and then continue to complete the checkout.

niranjan-uma-shankar commented:
> I'll just leave the In Progress label on for a while till the e2e tests pass. It makes sense to continue the review when we have all greens.

E2E tests pass.

> Until that gets fixed, we needs to test the with-theme flow by turning off the feature flag for customer home. 

I've a fix for this in https://github.com/Automattic/wp-calypso/pull/35649. I've also updated the testing instructions to include a new test for the with-theme flow.

@spen marking this PR ready for review again.

simison commented on client/my-sites/checklist/main.jsx:
```diff
@@ -97,6 +105,27 @@ class ChecklistMain extends PureComponent {
+								<a
+									href={ this.props.detailsUrl }
+									onClick={ () => this.trackClick( 'theme info', 'click' ) }
+								/>
```
Would it make sense to try to avoid having a link to the theme here, so to avoid customers getting distracted from finishing the checklist?

niranjan-uma-shankar replied:
Good point, I've removed the link in [5e7434d](https://github.com/Automattic/wp-calypso/pull/35283/commits/5e7434d4a9617959ff64efd18c73ba8986d875c9)

niranjan-uma-shankar committed:
Remove link to the theme details to avoid distracting from finishing the checklist

niranjan-uma-shankar commented:
Rebased after merging of #35649.

niranjan-uma-shankar committed:
Removing install=all

simison commented on client/my-sites/checkout/checkout/index.jsx:
```diff
@@ -394,7 +394,7 @@ export class Checkout extends React.Component {
 		if ( selectedSiteSlug && ( ! isReceiptEmpty || ! isCartEmpty ) ) {
 			if ( isJetpackNotAtomic ) {
-				return `/plans/my-plan/${ selectedSiteSlug }?thank-you&install=all`;
+				return `/plans/my-plan/${ selectedSiteSlug }?thank-you`;
```
This change we had to revert actually (https://github.com/Automattic/wp-calypso/pull/35654) so would be ace if you can put it back for now. Thanks. :-)

niranjan-uma-shankar replied:
Reverted in [f9dd97f](https://github.com/Automattic/wp-calypso/pull/35283/commits/f9dd97fd9d62ecabf15c95080d663d7d8ac1e073).

niranjan-uma-shankar committed:
Revert change to remove install=all

p-jackson commented:
I had some trouble making sure I had the most up to date changes, but I think I'm testing against the latest code now. However a number of the scenarios aren't working the way I'd expect from the PR description.

These are working like the description:
- [x] Scenario 2
- [x] Scenario 4 - see a thank you modal over the theme screen, which appears to be what was discussed as the correct behaviour in the PR comments
  <img width="869" alt="scenario 4 post theme" src="https://user-images.githubusercontent.com/1500769/63830535-809a2400-c9c0-11e9-9c09-d81af23268b2.png">
- [x] Scenario 5
- [x] Scenario 6
- [x] Scenario 7 - destination is not respected, which is expected

But the others aren't.

**Scenario 1**

After going through the Quick Start screen I am *always* seeing the thank you page, which is followed by the preview of the new site.
Just confirming that this scenario should be tested with a brand new account, to contrast it with scenario 2?

**Scenario 3**

After the checkout I see the thank you screen, although after clicking through that I see the site preview with the launch success banner, which is the ultimate expected destination from #35284 
<img width="905" alt="thank you page after launch" src="https://user-images.githubusercontent.com/1500769/63830599-a6bfc400-c9c0-11e9-82e1-3c6821ba0b53.png">

**Scenario 8**

When I choose a free plan I land on the theme specific thank you modal (like scenario 4). On a paid plan I go to the thank you page which will then take me to the site preview. So not respecting the destination.

sirreal commented on client/my-sites/checkout/checkout/index.jsx:
```diff
@@ -464,96 +553,49 @@ export class Checkout extends React.Component {
+			} );
 
-		if ( this.props.isJetpackNotAtomic ) {
-			// @FIXME temporary fix for plans purcahsed via `/plans` or WP Admin for connected sites
```
Reverting #35084 👍

sirreal reviewed (approved):
I tested the following Jetpack flows specifically for the issue in https://github.com/Automattic/wp-calypso/issues/35068 and this seems to work well.

- Purchase plan during Jetpack connection
- Upgrade to paid plan for already connected Jetpack site.

I did not do any significant code review or other testing.

chriskmnds commented on client/state/selectors/is-eligible-for-signup-destination.js:
```diff
@@ -27,23 +13,19 @@ import { retrieveSignupDestination } from 'signup/utils';
+		return false;
+	}
+
+	if ( destination.includes( '/checklist/' ) ) {
```
Just curious - do these `includes` calls get transpiled - cause they are not supported in IE? 🤔

sirreal replied:
I believe that's covered by our setup. I use modern features without worrying about browsers support.

- [browserslist](https://github.com/Automattic/wp-calypso/blob/0ecebd1057518f31316043d3dfa844d442e3ede1/package.json#L12-L34)
- [@babel/env](https://github.com/Automattic/wp-calypso/blob/0ecebd1057518f31316043d3dfa844d442e3ede1/babel.config.js#L11-L20)
- [@babel/polyfill](https://github.com/Automattic/wp-calypso/blob/0ecebd1057518f31316043d3dfa844d442e3ede1/client/boot/polyfills.js#L5)

chriskmnds replied:
Thanks for the pointers! It might not be. I remember coming across this before, just not certain in what context. I guess the answer may be in the browers list (?): 

```
	"last 2 Chrome versions",
	"last 1 ChromeAndroid versions",
	"last 2 Firefox versions",
	"last 2 Safari versions",
	"last 2 iOS versions",
	"last 2 Edge versions",
	"last 2 Opera versions",
	"unreleased Chrome versions",
	"unreleased ChromeAndroid versions",
	"unreleased Firefox versions",
	"unreleased FirefoxAndroid versions",
	"unreleased Edge versions",
	"unreleased Opera versions"
```
So no "IE".

Sorry if spontaneous comment btw (learning 😀).

simison replied:
Calypso builds two sets of bundles, one for modern browsers and another for everything else.

The "everything else" gets IE in browserslist via this package config:

https://github.com/WordPress/gutenberg/blob/8a3f3f2659d65a2e017a07bfe8852ef5d61d69bb/packages/browserslist-config/index.js#L3-L4

chriskmnds replied:
Gotcha - guess I should have followed through to check the defaults! Thanks for clarifying. :-)

(Nice to not worry about any of these)

chriskmnds replied:
Just to carify - the "evergreen" container should be transpiling for `ie>=11` right?

simison replied:
I'm not sure about details there, @sgomes from PerfOps implemented it so he knows. :-)

chriskmnds replied:
> I'm not sure about details there, @sgomes from PerfOps implemented it so he knows. :-)

That's good - no need. :-)

sgomes replied:
You rang? 😄 

As @simison pointed out, we have two builds: `evergreen`, which has its own config in our `package.json`, and `fallback`, which uses the defaults. The defaults are inherited from `@wordpress/browserslist`, as @simison pointed out too.

At runtime, on the server side, we look at the useragent for incoming requests. If it's a string we recognise as being from a modern browser, and a recent version thereof, we serve it the evergreen files. Otherwise, we serve it the fallback files.

This way, IE11 and any browser that we don't recognise as being modern will always get the fallback bundle, and thus all of the browser support it has built in, namely ES5 transpilation and polyfills for missing features (both manually added polyfills and polyfills added by CoreJS). Only browsers that we know for sure are modern will be sent the leaner, evergreen build, that includes a lot less of this.

In general, this should all be pretty transparent while you're writing code. If the feature you're using is supported by CoreJS v2 (the version we're currently using, and which is mostly language feature-focused), or one of our [manually-added polyfills](https://github.com/Automattic/wp-calypso/blob/master/client/boot/polyfills.js), you should be good to go! 🙂

chriskmnds replied:
Oh, glad you are here then! Thanks for the generous explanation @sgomes. Very good to know. :-)

niranjan-uma-shankar commented:
@p-jackson 

> After going through the Quick Start screen I am always seeing the thank you page, which is followed by the preview of the new site.
Just confirming that this scenario should be tested with a brand new account, to contrast it with scenario 2?

Here's a recording of Scenario 1 - https://cloudup.com/cvtea4qG08w. After Quick Start purchase, I'm seeing that I'm taken to the signup destination. 
Yes, this scenario is to be tested with a brand new account.

> When I choose a free plan I land on the theme specific thank you modal (like scenario 4). On a paid plan I go to the thank you page which will then take me to the site preview. So not respecting the destination.

Here's a recording of with-theme flow on a free plan, taken to theme specific checklist - https://cloudup.com/cVKHCto4ErY. 

> After the checkout I see the thank you screen, although after clicking through that I see the site preview with the launch success banner, which is the ultimate expected destination from #35284

I'm waiting for the 30 minutes to cross on my test account, but I'm reasonably confident that it will work given I've tested this before. I'll report back when I complete testing.

Could you please try again and verify that you are on the latest branch?

p-jackson commented:
Ok, it looks like it could be a FF issue! Here's my recording of scenario 1 using FF not incognito with a brand new account: https://cld.wthms.co/fGICzX

When I use Chrome I see the behaviour in your recording.

I've temporarily changed FF's Browser Privacy - Content Blocking setting to Standard to check whether this was causing the issue, but it's still broken with the less aggressive setting 😕

p-jackson reviewed (approved):
Have confirmed that FF doesn't treat the `wpcom_signup_complete_destination` cookie the same way in production and on `calypso.localhost`, and that it works correctly when in the production environment.

I think this is good to go :shipit:
